huffman/src/bit_span.hpp

```diff
@@ -113,5 +115,57 @@ class bit_span : public std::ranges::view_interface<bit_span>
   {
     return iterator{*this, bit_offset_ + bit_size_};
   };
+
+  template <class T>
+  constexpr auto pop() -> T
+  {
+    // TODO
+    // assert byte aligned
+    T res;
+    std::memcpy(&res, data_, sizeof(T));
+    std::advance(data_, sizeof(T));
+    if (std::endian::native == std::endian::big) {
+      res = std::byteswap(res);
+    }
+    return res;
+  }
+
+  constexpr auto pop_8() -> std::uint8_t { return pop<std::uint8_t>(); }
+
+  constexpr auto pop_16() -> std::uint16_t { return pop<std::uint16_t>(); }
+
+  /// Consumes the given number of bits. Basically advances the start of the
+  /// view.
+  ///
+  /// @pre n <= std::ranges::size(this)
+  constexpr auto consume(std::size_t n) -> void
+  {
+    assert(n <= bit_size_);
+    bit_size_ -= n;
+    bit_offset_ += n % CHAR_BIT;
+    if (bit_offset_ == CHAR_BIT) {
+      bit_offset_ = 0;
+      n += CHAR_BIT;
```

Comment:
I haven't convinced myself this is correct yet, particularly for n > 8. Maybe some tests?

In addition, you could also add an assertion

// invariant
assert(bit_offset_ < CHAR_BIT)

Reply:
Added some tests and the invariant.